File: app/src/main/java/com/s95ammar/weeklyschedule/views/fragments/ScheduleViewerFragment.java
package com.s95ammar.weeklyschedule.views.fragments;

import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.constraint.ConstraintLayout;
import android.support.constraint.ConstraintSet;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.s95ammar.weeklyschedule.R;
import com.s95ammar.weeklyschedule.models.CategoriesList;
import com.s95ammar.weeklyschedule.models.Day;
import com.s95ammar.weeklyschedule.models.Event;
import com.s95ammar.weeklyschedule.models.ScheduleItem;
import com.s95ammar.weeklyschedule.models.SchedulesList;

import java.io.Serializable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.HashMap;

public class ScheduleViewerFragment extends Fragment implements View.OnClickListener {
    private static final String TAG = "ScheduleViewerFragment";
    public static final int VIEW = 0;
    public static final int EDIT = 1;
    private static final int TEXT_VIEW_HEADER_HEIGHT = 100;
    private static final int TEXT_SIZE = 20;
    private static final int TEXT_VIEWS_WIDTH = 300;
    private static final int PADDING = 8;
    private ScheduleEditor mListener;
    private ScheduleItem schedule;
    private ConstraintLayout layoutScheduleViewer;
    private ArrayList<TextView> textViewsDays;
    private ArrayList<TextView> textViewsHours;
    private HashMap<Event, TextView> hashMapEvents;

    public ScheduleViewerFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        Bundle bundle = getArguments();
        if (bundle != null && schedule == null) {
            Serializable object = bundle.getSerializable(ScheduleEditor.KEY_SCHEDULE);
            if (object instanceof ScheduleItem) {
                schedule = (ScheduleItem) object;
            }
        }
        return inflater.inflate(R.layout.fragment_schedule_viewer, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setUpLayout();
        setUpActionButtonListener();
    }

    private void setUpLayout() {
        getActivity().setTitle(schedule != null ? schedule.getName() : getString(R.string.title_active_schedule));
        getView().findViewById(R.id.textView_no_active_schedule).setVisibility(schedule != null ? View.GONE : View.VISIBLE);
        if (schedule != null) {
            showSchedule();
            setMode(getArguments().getInt(ScheduleEditor.KEY_MODE, VIEW));
        }
    }

    public void setMode(@Mode int mode) {
        mListener.setDoneCancelVisibility(mode == EDIT);
        mListener.setEditVisibility(mode == VIEW);
        getView().findViewById(R.id.button_add_event).setVisibility(mode == EDIT ? View.VISIBLE : View.GONE);
        setEventsTextViewsOnClickListeners(mode);
    }

    public void showSchedule() {
        layoutScheduleViewer = getView().findViewById(R.id.layout_schedule_viewer);
        prepareHeaderTextViews(textViewsHours = new ArrayList<>(), Day.TOTAL_HOURS, Day.getHoursStringArray());
        prepareHeaderTextViews(textViewsDays = new ArrayList<>(), ScheduleItem.WEEK_DAYS.length, ScheduleItem.WEEK_DAYS);
        prepareEventTextViews();
        connectTextViews();
    }

    private void prepareHeaderTextViews(ArrayList<TextView> textViews, int length, String[] stringArray) {
            for (int i = 0; i < length; i++) {
                TextView tv = getTableTextView();
                formatHeaderTextView(tv, stringArray[i]);
                textViews.add(tv);
                layoutScheduleViewer.addView(tv);
            }
    }

    private void prepareEventTextViews() {
        hashMapEvents = new HashMap<>();
        for (int i = 0; i < schedule.getDays().size(); i++) {
            Day day = schedule.getDays().get(i);
            for (int j = 0; j < day.getEvents().size(); j++) {
                Event event = day.getEvents().get(j);
                TextView tv = getTableTextView();
                formatEventTextView(tv, event);
                hashMapEvents.put(event, tv);
                layoutScheduleViewer.addView(tv);
            }
        }
    }

    private TextView getTableTextView() {
        TextView tv = new TextView(getActivity());
        tv.setId(View.generateViewId());
        tv.setMinWidth(TEXT_VIEWS_WIDTH);
        tv.setTextSize(TEXT_SIZE);
        tv.setBackground(getActivity().getDrawable(R.drawable.shape_rounded_rectangle).mutate());
        tv.setPaddingRelative(PADDING, PADDING, PADDING, PADDING);
        tv.setGravity(Gravity.CENTER);
        return tv;
    }

    private void formatHeaderTextView(TextView tv, String text) {
        tv.setMinHeight(TEXT_VIEW_HEADER_HEIGHT);
        tv.setText(text);
        tv.setTextColor(Color.BLACK);
        tv.getBackground().mutate().setTint(Color.GRAY);
    }

    private void formatEventTextView(TextView tv, Event event) {
        tv.setText(event.getName());
        tv.setTextColor(event.getCategory().getTextColor());
        tv.getBackground().mutate().setTint(event.getCategory().getFillColor());
        ConstraintLayout.LayoutParams layoutParams = new ConstraintLayout.LayoutParams(TEXT_VIEWS_WIDTH, 0);
        tv.setLayoutParams(layoutParams);
    }

    private void connectTextViews() {
        ConstraintSet constraintSet = new ConstraintSet();
        constraintSet.clone(layoutScheduleViewer);
        connectHeaderTextViews(textViewsHours, constraintSet, false);
        connectHeaderTextViews(textViewsDays, constraintSet, true);
        connectEventTextViews(constraintSet);
        constraintSet.applyTo(layoutScheduleViewer);

    }

    private void setEventsTextViewsOnClickListeners(@Mode int mode) {
        for (int i = 0; i < schedule.getDays().size(); i++) {
            Day day = schedule.getDays().get(i);
            for (int j = 0; j < day.getEvents().size(); j++) {
                Event event = day.getEvents().get(j);
                TextView textView = hashMapEvents.get(event);
                switch (mode) {
                    case VIEW:
                        textView.setOnClickListener(v -> {});
                        break;
                    case EDIT:
                        textView.setOnClickListener(v ->
                                mListener.startEventRefactorActivity(event, SchedulesList.getInstance().indexOf(schedule)));
                        break;
                }
            }
        }
    }

    private void connectHeaderTextViews(ArrayList<TextView> arrayList, ConstraintSet constraintSet, boolean horizontal) {
        for (int i = 0; i < arrayList.size(); i++) {
            switch (i) {
                case 0:
                    constraintSet.connect(arrayList.get(i).getId(),
                            horizontal ? ConstraintSet.LEFT : ConstraintSet.TOP,
                            R.id.layout_schedule_viewer,
                            horizontal ? ConstraintSet.LEFT : ConstraintSet.TOP,
                            horizontal ? 300 : 100);
                    break;
                default:
                    constraintSet.connect(arrayList.get(i).getId(),
                            horizontal ? ConstraintSet.LEFT : ConstraintSet.TOP,
                            arrayList.get(i - 1).getId(),
                            horizontal ? ConstraintSet.RIGHT : ConstraintSet.BOTTOM);
                    break;
            }
        }
    }

    private void connectEventTextViews(ConstraintSet constraintSet) {
        for (int i = 0; i < schedule.getDays().size(); i++) {
            Day day = schedule.getDays().get(i);
            for (int j = 0; j < day.getEvents().size(); j++) {
                Event event = day.getEvents().get(j);
                TextView eventTextView = hashMapEvents.get(event);
                TextView targetDay = textViewsDays.get(schedule.getIndexOfDay(event.getDay()));
                TextView targetStartHour = textViewsHours.get(event.getStartTime().getHourOfDay());
                TextView targetEndHour = textViewsHours.get(event.getEndTime().getHourOfDay());
                double headerHeight = TEXT_VIEW_HEADER_HEIGHT;
                constraintSet.connect(eventTextView.getId(), ConstraintSet.TOP, targetStartHour.getId(), ConstraintSet.TOP,
                        (int) ((headerHeight /Day.MINUTES_IN_HOUR) * event.getStartTime().getMinuteOfHour()));
                constraintSet.connect(eventTextView.getId(), ConstraintSet.BOTTOM, targetEndHour.getId(), ConstraintSet.BOTTOM,
                        (int) ((headerHeight /Day.MINUTES_IN_HOUR) * (Day.MINUTES_IN_HOUR - event.getEndTime().getMinuteOfHour())));
                constraintSet.connect(eventTextView.getId(), ConstraintSet.LEFT, targetDay.getId(), ConstraintSet.LEFT);
                constraintSet.connect(eventTextView.getId(), ConstraintSet.RIGHT, targetDay.getId(), ConstraintSet.RIGHT);
                constraintSet.constrainDefaultHeight(eventTextView.getId(), ConstraintSet.MATCH_CONSTRAINT_SPREAD);
            }
        }
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof ScheduleEditor) {
            mListener = (ScheduleEditor) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement ScheduleEditor");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener.setEditVisibility(false);
        mListener.setDoneCancelVisibility(false);
        mListener = null;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.button_add_event:
                if (!CategoriesList.getInstance().isEmpty()) {
                    mListener.startEventRefactorActivity(null, SchedulesList.getInstance().indexOf(schedule));
                } else {
                    Toast.makeText(getActivity(), R.string.category_list_empty_error, Toast.LENGTH_SHORT).show();
                }
                break;
        }
    }

    private void setUpActionButtonListener() {
        FloatingActionButton fab = getView().findViewById(R.id.button_add_event);
        fab.setOnClickListener(this);

    }


    public interface ScheduleEditor {
        void setEditVisibility(boolean visibility);
        void setDoneCancelVisibility(boolean visibility);
        void startEventRefactorActivity(Event event, int scheduleIndex);
        void openScheduleViewerFragment(ScheduleItem schedule, @Mode int mode, int fragContainerId);
        String KEY_MODE = "mode";
        String KEY_SCHEDULE = "schedule";
        String KEY_SCHEDULE_INDEX = "schedule index";
        String KEY_EVENT = "event";
    }


    @IntDef({EDIT, VIEW})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Mode {
    }


}

File: app/src/main/java/com/s95ammar/weeklyschedule/views/fragments/ScheduleViewerFragment.kt
package com.s95ammar.weeklyschedule.views.fragments

import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.graphics.drawable.RippleDrawable
import android.os.Bundle
import android.util.TypedValue
import android.view.*
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.TextView
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.constraintlayout.widget.ConstraintSet
import androidx.core.widget.TextViewCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.findNavController
import com.s95ammar.weeklyschedule.R
import com.s95ammar.weeklyschedule.di.TimePattern
import com.s95ammar.weeklyschedule.models.data.Category
import com.s95ammar.weeklyschedule.models.data.Event
import com.s95ammar.weeklyschedule.models.data.Schedule
import com.s95ammar.weeklyschedule.util.*
import com.s95ammar.weeklyschedule.viewModels.ScheduleViewerViewModel
import com.s95ammar.weeklyschedule.viewModels.SharedDataViewModel
import dagger.android.support.DaggerFragment
import kotlinx.android.synthetic.main.fragment_schedule_viewer.*
import javax.inject.Inject


class ScheduleViewerFragment : DaggerFragment() {

	@Inject lateinit var factory: ViewModelProvider.Factory
	private lateinit var viewModel: ScheduleViewerViewModel
	private lateinit var sharedDataViewModel: SharedDataViewModel
	private lateinit var scheduleToolbarMenu: Menu

	private var schedule: Schedule
		get() = viewModel.schedule
		set(value) { viewModel.schedule = value }
	private var events: List<Event>
		get() = viewModel.events
		set(value) { viewModel.events = value }

	private val argScheduleId
		get() = arguments?.getInt(resources.getString(R.string.key_schedule_id)) ?: 0

	companion object {
		private const val TEXT_VIEW_HEADER_HEIGHT = 100
		private const val MIN_TEXT_SIZE = 4
		private const val TEXT_SIZE = 20
		private const val TEXT_VIEWS_WIDTH = 300
	}

	@field: [Inject TimePattern] lateinit var timePattern: String
	private val textViewsDays = ArrayList<TextView>()
	private val textViewsHours = ArrayList<TextView>()
	private var mapEventsTextViews = HashMap<Event, TextView>()

	private enum class Direction { Horizontal, Vertical }

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
	                          savedInstanceState: Bundle?): View? {
		setHasOptionsMenu(true)
		return inflater.inflate(R.layout.fragment_schedule_viewer, container, false)
	}

	override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
		super.onCreateOptionsMenu(menu, inflater)
		inflater.inflate(R.menu.toolbar_menu, menu)
		scheduleToolbarMenu = menu
		viewModel.scheduleMode.observe(viewLifecycleOwner, getModeObserver())
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		when (item.itemId) {
			R.id.button_edit -> viewModel.setMode(ScheduleMode.EDIT)
			R.id.button_done -> viewModel.setMode(ScheduleMode.VIEW)
		}
		return false
	}

	override fun onActivityCreated(savedInstanceState: Bundle?) {
		super.onActivityCreated(savedInstanceState)
		viewModel = ViewModelProvider(this, factory).get(ScheduleViewerViewModel::class.java)
		sharedDataViewModel = ViewModelProvider(requireActivity()).get(SharedDataViewModel::class.java)
		viewModel.getAllCategories().safeFetch { sharedDataViewModel.allCategories = it }
		setMode()
		if (viewModel.scheduleMode.value != ScheduleMode.MISSING) {
			displaySchedule()
			sharedDataViewModel.actionBarTitle.observe(viewLifecycleOwner, Observer { setActionBarTitle(it) })
		}
	}

	private fun setMode() {
		viewModel.scheduleMode.value ?: viewModel.setMode(when (argScheduleId) {
			0 -> ScheduleMode.MISSING
			else -> ScheduleMode.VIEW
		})
	}

	private fun getModeObserver(): Observer<ScheduleMode> = Observer { mode ->
		setScheduleToolbarMenuMode(mode)
		refreshEventsBackground(mode)
		when (mode) {
			ScheduleMode.MISSING -> textView_no_active_schedule.visibility = VISIBLE
			ScheduleMode.VIEW -> button_add_event.visibility = GONE
			ScheduleMode.EDIT -> {
				button_add_event.visibility = VISIBLE
				button_add_event.setOnClickListener {
					if (sharedDataViewModel.allCategories.isNotEmpty()) navigateToEventEditorFragment()
					else toast(R.string.category_list_empty_error, Toast.LENGTH_LONG)
				}
			}
		}
	}

	private fun setScheduleToolbarMenuMode(mode: ScheduleMode) {
		scheduleToolbarMenu.findItem(R.id.button_edit).isVisible = when (mode) {
			ScheduleMode.VIEW -> true
			else -> false
		}
		scheduleToolbarMenu.findItem(R.id.button_done).isVisible = when (mode) {
			ScheduleMode.EDIT -> true
			else -> false
		}
	}

	private fun refreshEventsBackground(mode: ScheduleMode) {
		mapEventsTextViews.forEach { (event, tv) ->
			sharedDataViewModel.allCategories.find { it.id == event.categoryId }?.let { category ->
				tv.background = getEventBackground(category, mode)
			}
		}
	}

	private fun getEventBackground(category: Category, mode: ScheduleMode): Drawable =
			getStrokedBackground(category.fillColor, category.textColor, rippleEffect = (mode == ScheduleMode.EDIT))

	private fun navigateToEventEditorFragment(event: Event? = null) {
		sharedDataViewModel.editedSchedule = schedule
		event?.let { sharedDataViewModel.editedEvent = event }
		sharedDataViewModel.eventEditorFragmentMode = event?.let { Mode.EDIT } ?: Mode.ADD
		findNavController().navigate(R.id.action_nav_schedule_viewer_to_eventEditorFragment)
	}

	private fun displaySchedule() {
		progressBar.visibility = VISIBLE
		viewModel.getScheduleById(argScheduleId).safeFetch {
			schedule = it
			sharedDataViewModel.setActionBarTitle(schedule.name)
			viewModel.getEventsBy(schedule.id).observe(viewLifecycleOwner, Observer { scheduleEvents ->
				progressBar.visibility = GONE
				prepareHeaderTextViews(textViewsHours, HOURS_IN_DAY, getHoursStringArray(timePattern))
				prepareHeaderTextViews(textViewsDays, schedule.days.amount, schedule.days.array)
				events = scheduleEvents
				prepareEventTextViews()
				connectTextViews()
			})
		}
	}


	private fun prepareHeaderTextViews(textViews: ArrayList<TextView>, length: Int, values: Array<String>) {
		textViews.clear()
		for (i in 0 until length) {
			val tv = getTableTextView()
			formatHeaderTextView(tv, values[i])
			textViews.add(tv)
			layout_schedule_viewer.addView(tv)
		}
	}

	private fun prepareEventTextViews() {
		mapEventsTextViews.clear()
		events.forEach { event ->
			val tv = getTableTextView()
			formatEventTextView(tv, event)
			mapEventsTextViews[event] = tv
			layout_schedule_viewer.addView(tv)
			tv.setOnClickListener {
				when (viewModel.scheduleMode.value) {
					ScheduleMode.EDIT -> navigateToEventEditorFragment(event)
					else -> {}
				}
			}

		}
	}

	private fun getTableTextView() = TextView(activity).apply {
		id = View.generateViewId()
		minWidth = TEXT_VIEWS_WIDTH
		TextViewCompat.setAutoSizeTextTypeUniformWithConfiguration(
				this, MIN_TEXT_SIZE, TEXT_SIZE, 1, TypedValue.COMPLEX_UNIT_SP
		)
		background = requireActivity().getDrawable(R.drawable.shape_rounded_rectangle)
		gravity = Gravity.CENTER
	}

	private fun formatHeaderTextView(tv: TextView, text: String) {
		tv.apply {
			minHeight = TEXT_VIEW_HEADER_HEIGHT
			this.text = text
			setTextColor(COLOR_BLACK)
			background.setTint(COLOR_GRAY)
		}
	}

	private fun formatEventTextView(tv: TextView, event: Event) {
		tv.apply {
			text = event.name
			viewModel.scheduleMode.value?.let { mode ->
				sharedDataViewModel.allCategories.find { it.id == event.categoryId }?.let { category ->
					tv.setTextColor(category.textColor)
					background = getEventBackground(category, mode)
				}
			}
			layoutParams = ConstraintLayout.LayoutParams(TEXT_VIEWS_WIDTH, 0)
		}
	}

	private fun connectTextViews() {
		val constraintSet = ConstraintSet()
		constraintSet.clone(layout_schedule_viewer)
		connectHeaderTextViews(textViewsHours, constraintSet, Direction.Vertical)
		connectHeaderTextViews(textViewsDays, constraintSet, Direction.Horizontal)
		connectEventTextViews(constraintSet)
		constraintSet.applyTo(layout_schedule_viewer)
	}

	private fun connectHeaderTextViews(headerTextViews: ArrayList<TextView>, constraintSet: ConstraintSet, direction: Direction) {
		val headerStartSide: Int
		val headerEndSide: Int
		val layoutEndSide: Int
		val marginToLayout: Int
		val marginInBetween = 2

		when (direction) {
			Direction.Horizontal -> {
				headerStartSide = ConstraintSet.LEFT
				headerEndSide = ConstraintSet.RIGHT
				layoutEndSide = ConstraintSet.LEFT
				marginToLayout = 300
			}
			Direction.Vertical -> {
				headerStartSide = ConstraintSet.TOP
				headerEndSide = ConstraintSet.BOTTOM
				layoutEndSide = ConstraintSet.TOP
				marginToLayout = 100
			}
		}

		for (i in headerTextViews.indices)
			when (i) {
				0 -> constraintSet.connect(headerTextViews[i].id, headerStartSide, layout_schedule_viewer.id, layoutEndSide, marginToLayout)
				else -> constraintSet.connect(headerTextViews[i].id, headerStartSide, headerTextViews[i - 1].id, headerEndSide, marginInBetween)
			}
	}

	private fun connectEventTextViews(constraintSet: ConstraintSet) {
		events.forEach { event ->
			mapEventsTextViews[event]?.id?.let { eventTextViewId ->
				val ec = EventConstraints(event)
				constraintSet.apply {
					connect(eventTextViewId, ConstraintSet.TOP, ec.targetStartHourId, ConstraintSet.TOP, ec.marginTop)
					connect(eventTextViewId, ConstraintSet.BOTTOM, ec.targetEndHourId, ConstraintSet.BOTTOM, ec.marginBottom)
					connect(eventTextViewId, ConstraintSet.LEFT, ec.targetDayId, ConstraintSet.LEFT)
					connect(eventTextViewId, ConstraintSet.RIGHT, ec.targetDayId, ConstraintSet.RIGHT)
					constrainDefaultHeight(eventTextViewId, ConstraintSet.MATCH_CONSTRAINT_SPREAD)
				}
			}
		}
	}

	private inner class EventConstraints(event: Event) {
		private val heightPerMinute = TEXT_VIEW_HEADER_HEIGHT.toDouble() / MINUTES_IN_HOUR

		val targetDayId: Int = textViewsDays[schedule.days.array.indexOf(event.day)].id
		val targetStartHourId: Int = textViewsHours[event.startHour].id
		val targetEndHourId: Int = textViewsHours[event.endHour].id

		val marginTop: Int = (heightPerMinute * event.startMinute).toInt()
		val marginBottom: Int = (heightPerMinute * (MINUTES_IN_HOUR - event.endMinute)).toInt()
	}

}
